File: RDMPEG/RDMPEGPlayer/RDMPEGFramebuffer/RDMPEGFramebuffer.m
//
//  RDMPEGFramebuffer.m
//  RDMPEG
//
//  Created by Serhii Alpieiev on 8/17/17.
//  Copyright © 2017 Readdle. All rights reserved.
//

#import "RDMPEGFramebuffer.h"
#import <Log4Cocoa/Log4Cocoa.h>



NS_ASSUME_NONNULL_BEGIN

@interface RDMPEGFramebuffer ()

@property (nonatomic, strong) NSMutableArray<RDMPEGVideoFrame *> *videoFrames;
@property (nonatomic, strong) NSMutableArray<RDMPEGAudioFrame *> *audioFrames;
@property (nonatomic, strong) NSMutableArray<RDMPEGSubtitleFrame *> *subtitleFrames;
@property (atomic, strong, nullable) RDMPEGArtworkFrame *artworkFrame;

@end



@implementation RDMPEGFramebuffer

#pragma mark - Overridden Class Methods

+ (L4Logger *)l4Logger {
    return [L4Logger loggerForName:@"rd.mediaplayer.RDMPEGFramebuffer"];
}

#pragma mark - Lifecycle

- (instancetype)init {
    self = [super init];
    if (self) {
        self.videoFrames = [NSMutableArray array];
        self.audioFrames = [NSMutableArray array];
        self.subtitleFrames = [NSMutableArray array];
    }
    return self;
}

#pragma mark - Public Accessors

- (NSTimeInterval)bufferedVideoDuration {
    @synchronized (self.videoFrames) {
        NSTimeInterval bufferedVideoDuration = 0.0;
        for (RDMPEGVideoFrame *videoFrame in self.videoFrames) {
            bufferedVideoDuration += videoFrame.duration;
        }
        return bufferedVideoDuration;
    }
}

- (NSTimeInterval)bufferedAudioDuration {
    @synchronized (self.audioFrames) {
        NSTimeInterval bufferedAudioDuration = 0.0;
        for (RDMPEGAudioFrame *audioFrame in self.audioFrames) {
            bufferedAudioDuration += audioFrame.duration;
        }
        return bufferedAudioDuration;
    }
}

- (NSTimeInterval)bufferedSubtitleDuration {
    @synchronized (self.subtitleFrames) {
        NSTimeInterval minSubtitlePosition = self.nextSubtitleFrame.position;
        NSTimeInterval maxSubtitlePosition = self.nextSubtitleFrame.position + self.nextSubtitleFrame.duration;
        
        for (RDMPEGSubtitleFrame *subtitleFrame in self.subtitleFrames) {
            minSubtitlePosition = MIN(minSubtitlePosition, subtitleFrame.position);
            maxSubtitlePosition = MAX(maxSubtitlePosition, subtitleFrame.position + subtitleFrame.duration);
        }
        
        return maxSubtitlePosition - minSubtitlePosition;
    }
}

- (NSUInteger)bufferedVideoFramesCount {
    @synchronized (self.videoFrames) {
        return self.videoFrames.count;
    }
}

- (NSUInteger)bufferedAudioFramesCount {
    @synchronized (self.audioFrames) {
        return self.audioFrames.count;
    }
}

- (NSUInteger)bufferedSubtitleFramesCount {
    @synchronized (self.subtitleFrames) {
        return self.subtitleFrames.count;
    }
}

- (nullable RDMPEGVideoFrame *)nextVideoFrame {
    @synchronized (self.videoFrames) {
        return self.videoFrames.firstObject;
    }
}

- (nullable RDMPEGAudioFrame *)nextAudioFrame {
    @synchronized (self.audioFrames) {
        return self.audioFrames.firstObject;
    }
}

- (nullable RDMPEGSubtitleFrame *)nextSubtitleFrame {
    @synchronized (self.subtitleFrames) {
        return self.subtitleFrames.firstObject;
    }
}

#pragma mark - Public Methods

- (void)pushFrames:(NSArray<RDMPEGFrame *> *)frames {
    for (RDMPEGFrame *frame in frames) {
        switch (frame.type) {
            case RDMPEGFrameTypeVideo: {
                RDMPEGVideoFrame *videoFrame = (RDMPEGVideoFrame *)frame;
                
#if defined(RD_DEBUG_MPEG_PLAYER)
                log4Debug(@"Pushed video frame: %f %f", frame.position, frame.duration);
#endif // RD_DEBUG_MPEG_PLAYER
                
                @synchronized (self.videoFrames) {
                    [self.videoFrames addObject:videoFrame];
                }
                break;
            }
            case RDMPEGFrameTypeAudio: {
                RDMPEGAudioFrame *audioFrame = (RDMPEGAudioFrame *)frame;
                
#if defined(RD_DEBUG_MPEG_PLAYER)
                log4Debug(@"Pushed audio frame: %f %f", frame.position, frame.duration);
#endif // RD_DEBUG_MPEG_PLAYER
                
                @synchronized (self.audioFrames) {
                    [self.audioFrames addObject:audioFrame];
                }
                break;
            }
            case RDMPEGFrameTypeSubtitle: {
                RDMPEGSubtitleFrame *subtitleFrame = (RDMPEGSubtitleFrame *)frame;
                
#if defined(RD_DEBUG_MPEG_PLAYER)
                log4Debug(@"Pushed subtitle frame: %f %f %@", subtitleFrame.position, subtitleFrame.duration, subtitleFrame.text);
#endif // RD_DEBUG_MPEG_PLAYER
                
                @synchronized (self.subtitleFrames) {
                    [self.subtitleFrames addObject:subtitleFrame];
                }
                break;
            }
            case RDMPEGFrameTypeArtwork: {
#if defined(RD_DEBUG_MPEG_PLAYER)
                log4Debug(@"Pushed artwork frame: %f %f", frame.position, frame.duration);
#endif // RD_DEBUG_MPEG_PLAYER
                
                self.artworkFrame = (RDMPEGArtworkFrame *)frame;
                break;
            }
        }
    }
}

- (nullable RDMPEGVideoFrame *)popVideoFrame {
    @synchronized (self.videoFrames) {
        RDMPEGVideoFrame *videoFrame = self.videoFrames.firstObject;
        
        if (videoFrame == nil) {
            return nil;
        }
        
        [self.videoFrames removeObjectAtIndex:0];
        
        return videoFrame;
    }
}

- (nullable RDMPEGAudioFrame *)popAudioFrame {
    @synchronized (self.audioFrames) {
        RDMPEGAudioFrame *audioFrame = self.audioFrames.firstObject;
        
        if (audioFrame == nil) {
            return nil;
        }
        
        [self.audioFrames removeObjectAtIndex:0];
        
        return audioFrame;
    }
}

- (nullable RDMPEGSubtitleFrame *)popSubtitleFrame {
    @synchronized (self.subtitleFrames) {
        RDMPEGSubtitleFrame *subtitleFrame = self.subtitleFrames.firstObject;
        
        if (subtitleFrame == nil) {
            return nil;
        }
        
        [self.subtitleFrames removeObjectAtIndex:0];
        
        return subtitleFrame;
    }
}

- (void)atomicVideoFramesAccess:(void (^)())accessBlock {
    if (accessBlock) {
        @synchronized (self.videoFrames) {
            accessBlock();
        }
    }
}

- (void)atomicAudioFramesAccess:(void (^)())accessBlock {
    if (accessBlock) {
        @synchronized (self.audioFrames) {
            accessBlock();
        }
    }
}

- (void)atomicSubtitleFramesAccess:(void (^)())accessBlock {
    if (accessBlock) {
        @synchronized (self.subtitleFrames) {
            accessBlock();
        }
    }
}

- (void)purge {
    [self purgeVideoFrames];
    [self purgeAudioFrames];
    [self purgeSubtitleFrames];
    [self purgeArtworkFrame];
}

- (void)purgeVideoFrames {
    @synchronized (self.videoFrames) {
        [self.videoFrames removeAllObjects];
    }
}

- (void)purgeAudioFrames {
    @synchronized (self.audioFrames) {
        [self.audioFrames removeAllObjects];
    }
}

- (void)purgeSubtitleFrames {
    @synchronized (self.subtitleFrames) {
        [self.subtitleFrames removeAllObjects];
    }
}

- (void)purgeArtworkFrame {
    self.artworkFrame = nil;
}

@end

NS_ASSUME_NONNULL_END

File: RDMPEG/RDMPEGPlayer/RDMPEGFramebuffer/RDMPEGFramebuffer.swift
//
//  RDMPEGFramebuffer.swift
//  RDMPEG
//
//  Created by Max Berezhnoy on 15/07/2024.
//  Copyright © 2024 Readdle. All rights reserved.
//

import Foundation
import Log4Cocoa

class RDMPEGFramebuffer {
    private(set) var artworkFrame: RDMPEGArtworkFrame?

    private var videoFrames: [RDMPEGVideoFrame] = []
    private var audioFrames: [RDMPEGAudioFrame] = []
    private var subtitleFrames: [RDMPEGSubtitleFrame] = []
    private var videoFramesLock = NSRecursiveLock()
    private var audioFramesLock = NSRecursiveLock()
    private var subtitleFramesLock = NSRecursiveLock()

    var bufferedVideoDuration: TimeInterval {
        videoFramesLock.withLock {
            return videoFrames.reduce(0) { $0 + $1.duration }
        }
    }

    var bufferedAudioDuration: TimeInterval {
        audioFramesLock.withLock {
            return audioFrames.reduce(0) { $0 + $1.duration }
        }
    }

    var bufferedSubtitleDuration: TimeInterval {
        subtitleFramesLock.withLock {
            guard let first = subtitleFrames.first else { return 0 }
            var minPosition = first.position
            var maxPosition = first.position + first.duration

            for frame in subtitleFrames {
                minPosition = min(minPosition, frame.position)
                maxPosition = max(maxPosition, frame.position + frame.duration)
            }

            return maxPosition - minPosition
        }
    }

    var bufferedVideoFramesCount: Int {
        videoFramesLock.withLock {
            return videoFrames.count
        }
    }

    var bufferedAudioFramesCount: Int {
        audioFramesLock.withLock {
            return audioFrames.count
        }
    }

    var bufferedSubtitleFramesCount: Int {
        subtitleFramesLock.withLock {
            return subtitleFrames.count
        }
    }

    var nextVideoFrame: RDMPEGVideoFrame? {
        videoFramesLock.withLock {
            return videoFrames.first
        }
    }

    var nextAudioFrame: RDMPEGAudioFrame? {
        audioFramesLock.withLock {
            return audioFrames.first
        }
    }

    var nextSubtitleFrame: RDMPEGSubtitleFrame? {
        subtitleFramesLock.withLock {
            return subtitleFrames.first
        }
    }

    class var l4Logger: L4Logger {
        return L4Logger(forName: "rd.mediaplayer.RDMPEGFramebuffer")
    }

    func pushFrames(_ frames: [RDMPEGFrame]) {
        for frame in frames {
            switch frame.type {
            case .video:
                if let videoFrame = frame as? RDMPEGVideoFrame {
                    #if RD_DEBUG_MPEG_PLAYER
                    log4Debug("Pushed video frame: \(frame.position) \(frame.duration)")
                    #endif
                    videoFramesLock.withLock {
                        videoFrames.append(videoFrame)
                    }
                }
            case .audio:
                if let audioFrame = frame as? RDMPEGAudioFrame {
                    #if RD_DEBUG_MPEG_PLAYER
                    log4Debug("Pushed audio frame: \(frame.position) \(frame.duration)")
                    #endif
                    audioFramesLock.withLock {
                        audioFrames.append(audioFrame)
                    }
                }
            case .subtitle:
                if let subtitleFrame = frame as? RDMPEGSubtitleFrame {
                    #if RD_DEBUG_MPEG_PLAYER
                    log4Debug("""
                        Pushed subtitle frame:
                        \(subtitleFrame.position) \(subtitleFrame.duration) \(subtitleFrame.text ?? "")
                    """)
                    #endif
                    subtitleFramesLock.withLock {
                        subtitleFrames.append(subtitleFrame)
                    }
                }
            case .artwork:
                #if RD_DEBUG_MPEG_PLAYER
                log4Debug("Pushed artwork frame: \(frame.position) \(frame.duration)")
                #endif
                artworkFrame = frame as? RDMPEGArtworkFrame
            @unknown default:
                break
            }
        }
    }

    func popVideoFrame() -> RDMPEGVideoFrame? {
        videoFramesLock.withLock {
            guard !videoFrames.isEmpty else { return nil }
            return videoFrames.removeFirst()
        }
    }

    func popAudioFrame() -> RDMPEGAudioFrame? {
        audioFramesLock.withLock {
            guard !audioFrames.isEmpty else { return nil }
            return audioFrames.removeFirst()
        }
    }

    @discardableResult
    public func popSubtitleFrame() -> RDMPEGSubtitleFrame? {
        subtitleFramesLock.withLock {
            guard !subtitleFrames.isEmpty else { return nil }
            return subtitleFrames.removeFirst()
        }
    }

    func atomicVideoFramesAccess(_ accessBlock: () -> Void) {
        videoFramesLock.withLock {
            accessBlock()
        }
    }

    func atomicAudioFramesAccess(_ accessBlock: () -> Void) {
        audioFramesLock.withLock {
            accessBlock()
        }
    }

    func atomicSubtitleFramesAccess(_ accessBlock: () -> Void) {
        subtitleFramesLock.withLock {
            accessBlock()
        }
    }

    func purge() {
        purgeVideoFrames()
        purgeAudioFrames()
        purgeSubtitleFrames()
        purgeArtworkFrame()
    }

    func purgeVideoFrames() {
        videoFramesLock.withLock {
            videoFrames.removeAll()
        }
    }

    func purgeAudioFrames() {
        audioFramesLock.withLock {
            audioFrames.removeAll()
        }
    }

    func purgeSubtitleFrames() {
        subtitleFramesLock.withLock {
            subtitleFrames.removeAll()
        }
    }

    func purgeArtworkFrame() {
        artworkFrame = nil
    }
}
